nni/retiarii/oneshot/pytorch/base_lightning.py

```diff
@@ -0,0 +1,223 @@
+    nas_modules : list[nn.Module]
+        modules a paticular NAS method replaces with
+
+    Parameters
+    ----------
+    base_model : pl.LightningModule
+        The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+        is wrapped by base_model, and base_model will be wrapped by this model.
+    custom_replace_dict : Dict[Type[nn.Module], Callable[[nn.Module], nn.Module]], default = None
+        The custom xxxChoice replace method. Keys should be xxxChoice type and values should
+        return an nn.module. This custom replace dict will override the default replace
+        dict of each NAS method.
+    """
+    def __init__(self, base_model, custom_replace_dict=None):
+        super().__init__()
+        assert isinstance(base_model, pl.LightningModule)
+        self.model = base_model
+
+        # replace xxxChoice with respect to NAS alg
+        # replaced modules are stored in self.nas_modules
+        self.nas_modules = []
+        choice_replace_dict = self.default_replace_dict
+        if custom_replace_dict is not None:
+            for k, v in custom_replace_dict.items():
+                assert isinstance(v, nn.Module)
+                choice_replace_dict[k] = v
+        _replace_module_with_type(
+            self.model, choice_replace_dict, self.nas_modules)
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        # You can use self.optimizers() in training_step to get a list of all optimizers.
+        # Model optimizers comes after architecture optimizers, and the number of architecture
+        # optimizers is self.arc_optim_count.
+        #
+        # Example :
+        # optims = self.optimizers()
+        # arc_optims = optims[:self.arc_optim_count] # architecture optimizers
+        # w_optimizers = optims[self.arc_optim_count:] # model optimizers
+
+        return self.model.training_step(batch, batch_idx)
+
+    def validation_step(self, batch, batch_idx):
+        warn('Validation is skipped by the NAS method you chose.')
+
+    def configure_optimizers(self):
+        """
+        Combine architecture optimizers and user's model optimizers.
+        Overwrite configure_architecture_optimizers if architecture optimizers
+        are needed in your NAS algorithm.
+        By default ''self.model'' is currently a _SupervisedLearningModule in
+        nni.retiarii.evaluator.pytorch.lightning, and it only returns 1 optimizer.
+        But for extendibility, codes for other return value types are also implemented.
+        """
+        arc_optimizers = self.configure_architecture_optimizers()
+        if arc_optimizers is None:
+            return self.model.configure_optimizers()
+
+        if isinstance(arc_optimizers, optim.Optimizer):
+            arc_optimizers = [arc_optimizers]
+        self.arc_optim_count = len(arc_optimizers)
+
+        # The third return value ''frequency'' and ''monitor'' are ignored since lightning
```

Comment:
I think `lr_schedulers` is not automatically called when manual optimization is on. https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
https://pytorch-lightning.readthedocs.io/en/stable/common/optimizers.html#learning-rate-scheduling-manual

Please check whether we need to call `lr_scheduler.step()` either in training step or in training epoch end.